message_tagging_service/tagging_service.py

```diff
@@ -259,17 +261,46 @@ def match(self, modulemd):
                         self._property_matches.append(False)
 
         if all(self._property_matches):
-            if self._regex_has_named_group:
-                formatted_dest_tags = [
-                    re.sub(regex, self.destinations, mmd_property_value)
-                    for regex, mmd_property_value in self._regex_has_named_group
-                ]
-                return RuleMatch(True, formatted_dest_tags)
+            if self._regex_group_dicts:
+                return RuleMatch(True, self._generate_destination_tags())
             else:
                 return RuleMatch(True, [self.destinations])
         else:
             return RuleMatch(False)
 
+    def _generate_destination_tags(self):
+        # In some cases, the destination tag template uses multiple regex groups, and
+        # the value for these regex groups are extracted from different attributes of
+        # the modulemd info. As such a simple, re.sub of every match will throw an
+        # error because not all regex groups will be resolved. This method adds the
+        # ability to support this use case by generating permutations of each
+        # possible combination and applying each of these permutations to generate
+        # the destination tag name.
+
+        # Aggregate the different values for each regex group.
+        replacements = {}
+        for group_dict in self._regex_group_dicts:
+            for group, value in group_dict.items():
+                replacements.setdefault(group, [])
+                replacements[group].append(value)
```

Comment:
Why need to consider the case for example `{'stream': ['8.1', ...]}`?

Reply:
This is needed so it can generate multiple tags. Without this, existing test cases fail.